### PR TITLE
Add event for ticking players

### PR DESCRIPTION
**Pull request by Boundarybreaker**
I think it's a good idea to have something to tick equipped items, so that they can give players various effects like the turtle helmet, along with anything else equipment might want to do. I've designed it in a  way that's extensible for Baubles-likes, where it's passed a `Player`, an `Inventory`, an `int`, and an `ItemStack` so that mods can know what slot something is in. The current mixin for the event only targets vanilla equipment slots, and has to do so a bit weirdly due to how PlayerInventory is set up.

### CODE REVIEW

**Comment by Pyrofab** [comment]
If this is designed with extensibility in mind, maybe it should use an `Identifier` instead of an arbitrary number. The identifier for vanilla slots could be constructed from their name.

**Comment by Boundarybreaker** [comment]
hmm, that might work. I'd need to make a method for getting a slot from an Identifier and vice versa.

**Comment by Boundarybreaker** [comment]
Actually, I don't need to. The enum basically has that already.

File: src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(PlayerEntity.class)
+public abstract class MixinPlayerEntity extends LivingEntity {
+	protected MixinPlayerEntity(EntityType<? extends LivingEntity> type, World world) {
+		super(type, world);
+	}
+
+	@Inject(method = "updateTurtleHelmet", at = @At("HEAD"))
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: Any reason for injecting into `updateTurtleHelmet` wouldnt at the end of `update` be cleaner?

**Comment by Pyrofab** [comment]
It may still be worth it to try to avoid creating a new instance every time. Also, how would modded slots be ticked ?

File: src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java (hunk 1 above)
Comment (Boundarybreaker) [reply src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: Mainly for organization’s sake. It felt kinda wrong to me to not use the method that’s already there to tick equipment.

**Comment by Boundarybreaker** [comment]
@Pyrofab They’d add their own mixin to call the event for their slots.

**Comment by Pyrofab** [comment]
At this point, wouldn't it be simpler to just add a `PlayerTickCallback` that people can use however they want ?
```java
EquipmentTickCallback.EVENT.register((player, inv, slotId, stack) -> {
				if (stack.getItem() == MY_ITEM && slotId.equals(SLOT_ID)) doStuff();
			});
```
vs
```java
PlayerTickCallback.EVENT.register(player -> {
				if (player.getEquippedStack(SLOT).getItem() == MY_ITEM) doStuff();
			}
```

**Comment by raphydaphy** [comment]
We pretty much have that with client/server tick callback

**Comment by Pyrofab** [comment]
Except server tick callback is definitely more unwieldy if you just want to tick equipment slots.

**Comment by Boundarybreaker** [comment]
Honestly, switching to a generic player tick would probably be a lot better. I’ll do that when I have time; I’m flying home for spring break today so I’m not anywhere I can really do programming rn. I’ll also move it to `update()`’s tail like @modmuss50 recommended if it’s not gonna be equipment-exclusive.

**Comment by Pyrofab** [comment]
While you are at it you could add an example mod that uses the event to tick equipment.

**Comment by grondag** [comment]
This PR is labeled as an equipment tick, but it acts as a player tick event that issues an event callback for every slot, every tick.  If I understand it correctly, to tick my item I need to listen to this event and discard the vast majority of callbacks, testing each call to see if it is actually my item before executing my item tick logic.  So, two immediate observations:

1) I agree with the notion of a generic player tick event, not specific to equipment. That event shouldn't iterate slots - it should simply call back every listener for every player, every tick. It would be nice if listeners could optionally register for callbacks with a _specific_ player, but this isn't critical - the number of players is usually small.

2) A proper equipment tick could still be useful, but to do it in a performant way would be more complicated...

A way that comes to mind would be supporting the ```Tickable``` interface on items.  When items are added or removed from a player inventory (and at startup when the player is created/deserialized) the items are checked for this interface and we maintain a per-player collection of ```Tickable``` instances.  Ideally the collection includes slot - but that means all Inventory changes need to be hooked to maintain the collection.

All ```Tickable``` instances extant on the player are called each tick, with information about their current slot and of course the player instance itself.  No iteration of slots is necessary if the collection is maintained with slot info.

A simpler version would iterate each slot like the current PR does - but only once per tick - and only issue callbacks to items that implement the interface.  Less efficient but TBH probably good enough and much easier to implement. 

Supporting ```Tickable``` on items also suggests the idea that items might tick when held in containers if the container supports that capability and is loaded.  If we went that route, the player implementation would simply be a specific use case.

**Comment by Boundarybreaker** [comment]
Yeah, it’s being changed over to a generic player tick, I just can’t commit the change yet since no wifi.

**Comment by Boundarybreaker** [comment]
@modmuss50 changed to a player-specific event now.

File: src/main/java/net/fabricmc/fabric/api/event/player/PlayerTickCallback.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.minecraft.entity.EquipmentSlot;
+import net.minecraft.entity.player.PlayerEntity;
+
+/**
+ * Callback for ticking a player's equipment.
+ * Takes a player, the inventory to check, the slot ID, and the stack.
+ * Can be extended to use any Inventory for use by Baubles-likes. Don't always assume a PlayerInventory.
```
Comment (grondag) [inline src/main/java/net/fabricmc/fabric/api/event/player/PlayerTickCallback.java]: Java docs are no longer accurate

File: src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(PlayerEntity.class)
+public abstract class MixinPlayerEntity extends LivingEntity {
+	protected MixinPlayerEntity(EntityType<? extends LivingEntity> type, World world) {
+		super(type, world);
+	}
+
+	@Inject(method = "update", at = @At("TAIL"))
+	public void tickEquipmentSlots(CallbackInfo ci) {
```
Comment (grondag) [inline src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: Should this be called something other than ```tickEquipmentSlots``` now that this isn't an equipment-specific event?

Comment (Boundarybreaker) [reply src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: yes, it should. fixing.

**Comment by Boundarybreaker** [comment]
@modmuss50 can you review again?

**Comment by modmuss50** [comment]
Starting to look good, one thing that might be worth adding is support for profiling, see an example of this @ https://github.com/FabricMC/fabric/blob/master/src/main/java/net/fabricmc/fabric/api/event/client/ClientTickCallback.java#L26

**Comment by Boundarybreaker** [comment]
Can this be merged sometime soon? I've got a few things I wanna use this for.

**Comment by Boundarybreaker** [comment]
This has two approvals and is merge-ready. We should get it in before 1.14 full release.

**Comment by valoeghese** [comment]
I vote merge this

**Comment by InsomniaKitten** [comment]
This should be generalized to an Entity tick callback, imo. It's an expensive hook as-is, and if anyone were to ever want a tick callback for other entity types in the future, it would require duplicating this behaviour, essentially doubling execution time per tick.

File: src/main/java/net/fabricmc/fabric/api/event/player/PlayerTickCallback.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Callback for ticking a player. Useful for updating effects given by a player's equipped items.
+ */
+public interface PlayerTickCallback {
+	public static final Event<PlayerTickCallback> EVENT = EventFactory.createArrayBacked(PlayerTickCallback.class,
+			(listeners) -> {
+				if (EventFactory.isProfilingEnabled()) {
+					return (player) -> {
+						player.getServer().getProfiler().push("fabricPlayerTick");
```
Comment (InsomniaKitten) [inline src/main/java/net/fabricmc/fabric/api/event/player/PlayerTickCallback.java]: The call to `player.getServer().getProfiler()` should be stored in a local variable rather than being called multiple times within the loop of the callback.

```java
return (player) -> {
  Profiler profiler = player.getServer().getProfiler();
  profiler.push("fabricPlayerTick");
  for (PlayerTickCallback event : listeners) {
    profiler.push(EventFactory.getHandlerName(event));
    event.tick(player);
    profiler.pop();
  }
  profiler.pop();
};
```

**Comment by Pyrofab** [comment]
> 
> 
> This should be generalized to an Entity tick callback, imo. It's an expensive hook as-is, and if anyone were to ever want a tick callback for other entity types in the future, it would require duplicating this behaviour, essentially doubling execution time per tick.

You do have a point that adding an Entity tick callback later on would require duplicating. 
However, I disagree with the scale of the impact. This callback would be called once per player, per tick, which is negligible compared to once for every entity in the game (easily hundreds of time more).
Personally, I would not mind changing it to an entity callback, depends on peoples' needs.

File: src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+@Mixin(PlayerEntity.class)
+public abstract class MixinPlayerEntity extends LivingEntity {
+	protected MixinPlayerEntity(EntityType<? extends LivingEntity> type, World world) {
+		super(type, world);
+	}
+
+	@Inject(method = "update", at = @At("TAIL"))
+	public void tickPlayer(CallbackInfo ci) {
+		PlayerEntity player = (PlayerEntity)(Object)this;
```
Comment (InsomniaKitten) [inline src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: Creating a local variable here each tick is redundant, this can just be
```java
PlayerTickCallback.EVENT.invoker().tick((PlayerEntity) (Object) this);
```

File: src/test/java/net/fabricmc/fabric/events/PlayerTickEventMod.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import net.minecraft.text.TextComponent;
+
+
+public class PlayerTickEventMod implements ModInitializer {
+	@Override
+	public void onInitialize() {
+		PlayerTickCallback.EVENT.register(player -> {
+			ItemStack stack = player.getEquippedStack(EquipmentSlot.HEAD);
+			if (stack.getItem() == Items.PLAYER_HEAD) {
+				//if someone is wearing asie's head, make them glow
```
Comment (InsomniaKitten) [inline src/test/java/net/fabricmc/fabric/events/PlayerTickEventMod.java]: Tests should use java docs rather than line comments imo

File: src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.world.World;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(PlayerEntity.class)
+public abstract class MixinPlayerEntity extends LivingEntity {
```
Comment (InsomniaKitten) [inline src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: There is no reason to extend `LivingEntity` here, this can just be a superclass of `Object`

File: src/main/resources/net.fabricmc.fabric.mixins.common.json [hunk 8]
```diff
@@ -38,4 +39,4 @@
   "injectors": {
     "defaultRequire": 1
   }
-}
```
Comment (InsomniaKitten) [inline src/main/resources/net.fabricmc.fabric.mixins.common.json]: Unnecessary change

File: src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java (hunk 7 above)
Comment (Pyrofab) [reply src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: Except it is recommended to extend the target's superclass to avoid wasting time in hierarchy validation. (https://github.com/SpongePowered/Mixin/wiki/About-Hierarchy-Validation-in-Mixins)

Comment (InsomniaKitten) [reply src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: Huh.. my bad, I didn't realise class hierarchy was actually utilized here for validation. I should go and update my own mixins...

Comment (Pyrofab) [reply src/main/java/net/fabricmc/fabric/mixin/events/tick/MixinPlayerEntity.java]: To be fair, `Object` is the second most optimized thing to extend so it's not a big deal.

File: src/main/resources/net.fabricmc.fabric.mixins.common.json (hunk 8 above)
Comment (Boundarybreaker) [reply src/main/resources/net.fabricmc.fabric.mixins.common.json]: This is just git weirdness, really. Not much I can do about it.

File: src/test/java/net/fabricmc/fabric/events/PlayerTickEventMod.java (hunk 6 above)
Comment (Boundarybreaker) [reply src/test/java/net/fabricmc/fabric/events/PlayerTickEventMod.java]: That's out of scope of this PR, multiple other test classes already use comments instead of javadocs

File: src/main/resources/net.fabricmc.fabric.mixins.common.json (hunk 8 above)
Comment (InsomniaKitten) [reply src/main/resources/net.fabricmc.fabric.mixins.common.json]: You added a new line to the end of the file (probably ran an auto-format?)

Comment (Boundarybreaker) [reply src/main/resources/net.fabricmc.fabric.mixins.common.json]: Ahhh yeah, that'd be autoformat.

File: src/test/java/net/fabricmc/fabric/events/PlayerTickEventMod.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+				TextComponent name = stack.getItem().getTranslatedNameTrimmed(stack);
+				if (name.getText().equals("asiekierka")) {
```
Comment (asiekierka) [inline src/test/java/net/fabricmc/fabric/events/PlayerTickEventMod.java]: hah!

**Comment by Boundarybreaker** [comment]
This PR is outdated, and has been replaced by #184.
